source/server/server.cc

```diff
@@ -532,8 +533,13 @@ void InstanceImpl::shutdown() {
   ENVOY_LOG(info, "shutting down server instance");
   shutdown_ = true;
   restarter_.terminateParent();
-  notifyCallbacksForStage(Stage::ShutdownExit);
-  dispatcher_->exit();
+  auto completion_cb_count =
+      std::make_shared<std::atomic<int>>(stage_completable_callbacks_.size());
+  notifyCallbacksForStage(Stage::ShutdownExit, [this, completion_cb_count] {
```

Comment:
Would some form of assertion that this callback is executing on the right thread make sense?

Reply:
Added an assert.

Also refactored the counting logic into a wrapper callback created in the notifyCallbacksForStage method so it can be reused for other stages in the future. This also it makes the calling code easier to read.